tests/Analyze/Translate.hs

```diff
@@ -0,0 +1,282 @@
+
+    -- Like mkApp but for functions that take two arguments, the second of
+    -- which is a list. This pattern is used in `enforce-one` and `format`
+    mkApp'
+      :: NativeDef
+      -> ETerm
+      -> [ETerm]
+      -> ReaderT (GenEnv, GenState) Maybe (Pact.Term Pact.Ref)
+    mkApp' (_, defTm) arg argList = do
+      arg'     <- toPactTm arg
+      argList' <- traverse toPactTm argList
+      pure $ TApp (liftTerm defTm)
+        [arg', Pact.TList argList' (Pact.TyList Pact.TyAny) dummyInfo]
+        dummyInfo
+
+    arithOpToDef :: ArithOp -> NativeDef
+    arithOpToDef = \case
+      Add -> addDef
+      Sub -> subDef
+      Mul -> mulDef
+      Div -> divDef
+      Pow -> powDef
+      Log -> logDef
+
+    unaryArithOpToDef :: UnaryArithOp -> NativeDef
+    unaryArithOpToDef = \case
+      Negate -> subDef
+      Sqrt   -> sqrtDef
+      Ln     -> lnDef
+      Exp    -> expDef
+      Abs    -> absDef
+      Signum -> error "not yet implemented: we don't generate this operator"
+
+    roundingLikeOpToDef :: RoundingLikeOp -> NativeDef
+    roundingLikeOpToDef = \case
+      Round   -> roundDef
+      Ceiling -> ceilDef
+      Floor   -> floorDef
+
+    comparisonOpToDef :: ComparisonOp -> NativeDef
+    comparisonOpToDef = \case
+      Gt  -> gtDef
+      Lt  -> ltDef
+      Gte -> gteDef
+      Lte -> lteDef
+      Eq  -> eqDef
+      Neq -> neqDef
+
+    logicalOpToDef :: LogicalOp -> NativeDef
+    logicalOpToDef = \case
+      AndOp -> andDef
+      OrOp  -> orDef
+      NotOp -> notDef
+
+toPactTm'
+  -- :: Applicative m
+  :: Monad m
+  => (GenEnv, GenState) -> ETerm -> MaybeT m (Pact.Term Pact.Ref)
+toPactTm' envState etm = MaybeT $ do
+  pure $ runReaderT (toPactTm etm) envState
+
+toAnalyze :: Pact.Type (Pact.Term Pact.Ref) -> Pact.Term Pact.Ref -> MaybeT IO ETerm
+toAnalyze ty tm = do
+  let cnst = TConst
+        (Pact.Arg "tm" ty dummyInfo)
+        "module"
+        (Pact.CVRaw tm)
+        (Meta Nothing Nothing)
+        dummyInfo
+      ref = Pact.Ref cnst
+  maybeConst <- lift $ Pact.runTC 0 False $ typecheckTopLevel ref
+  (_cTy, ast) <- case maybeConst of
+    (Pact.TopConst _info _name constTy constAst _meta, _tcState)
+      -> pure (constTy, constAst)
+    _ -> MaybeT $ pure Nothing
+
+  -- TODO: this is all copied from Translate.hs
+  -- TODO: good chance we'll have to change some of these like the initial var
+  --       id
+  let vertex0 = 0
+      nextVertex = succ vertex0
+      path0 = Path 0
+      nextTagId = succ $ _pathTag path0
+      graph0 = pure vertex0
+      state0 = TranslateState nextTagId 0 graph0 vertex0 nextVertex Map.empty
+        mempty path0 Map.empty
+
+  hoist generalize $
+    exceptToMaybeT $
+      fmap fst $
+        flip runStateT state0 $
+          runReaderT
+            (unTranslateM (translateNode ast))
+            (mkTranslateEnv IsTest dummyInfo [])
+
+-- This is limited to simple types for now
+reverseTranslateType :: Type a -> Pact.Type b
```

Comment:
should we have an `Iso`?

Reply:
It's more like an adjunction: `maybeTranslateType :: Pact.Type Pact.UserType -> Maybe EType`